mongo_orchestration/sharded_clusters.py

```diff
@@ -68,23 +75,111 @@ def __init__(self, params):
                     {'$addToSet': {'$each': self.tags[sh_id]}})
 
         if self.login:
-            client = MongoClient(self.router['hostname'], **self.kwargs)
-            client.admin.add_user(self.login, self.password,
-                                  roles=['__system',
-                                         'clusterAdmin',
-                                         'dbAdminAnyDatabase',
-                                         'readWriteAnyDatabase',
-                                         'userAdminAnyDatabase'])
+            # Do we need to add an extra x509 user?
+            def only_x509(config):
+                set_params = config.get('setParameter', {})
+                auth_mechs = set_params.get('authenticationMechanisms', '')
+                auth_mechs = auth_mechs.split(',')
+                if len(auth_mechs) == 1 and auth_mechs[0] == 'MONGODB-X509':
+                    return True
+                return False
+
+            any_only_x509 = lambda l: any(map(only_x509, l))
+            shard_configs = [s.get('shardParams', {}).get('procParams', {})
+                             for s in params.get('shards', [])]
+            rs_shard_configs = [
+                m.get('procParams', {})
+                for s in params.get('shards', [])
+                for m in s.get('shardParams', {}).get('members', [])
+            ]
+            router_configs = params.get('routers', [])
+
+            self.x509_extra_user = (any_only_x509(configsvr_configs) or
+                                    any_only_x509(shard_configs) or
+                                    any_only_x509(rs_shard_configs) or
+                                    any_only_x509(router_configs))
+
+            self._add_users(self.connection()[self.auth_source])
+
+            # Secondary user given from request.
+            secondary_login = {
+                'name': self.login,
+                'writeConcern': {'w': 'majority'},
```

Comment:
As mentioned earlier, this won't work with servers before 2.6. You have to set write concern on the database instance.